### PR TITLE
DenseNet Model

### PR DESCRIPTION
**Pull request by forsyth2**
Add DenseNet model to LBANN's model zoo.

### CODE REVIEW

**Comment by timmoon10** [comment]
I've found this Python script to be helpful in generating complicated models:
https://github.com/timmoon10/lbann/blob/vram-dev/model_zoo/models/vram/generate_dram.py

File: model_zoo/models/densenet/generator_densenet.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  elif version == 161:
+    repeat = 24
+  dense_block(f, tab, num, repeat)
+  classification_layer(f, tab)
+  f.write('}\n')
+
+def dense_block(f, tab, num, repeat):
+  name = 'dense_block_%d' % num
+  for i	in range(repeat):
+    num = dense_conv(f, tab, name, i)
```
Comment (forsyth2) [inline model_zoo/models/densenet/generator_densenet.py]: remove assignment.

File: model_zoo/models/densenet/generator_densenet.py [hunk 2]
```diff
@@ -8,10 +8,32 @@ def dense_net(prototext_file='model_densenet.prototext', version=121):
   f = open(prototext_file, 'w')
   f.write('model {\n')
   tab = ' '*2
-  f.write(tab + 'name: "directed_acyclic_graph_model"\n')
-  f.write(tab + '###################################################\n')
-  f.write(tab + '# Layers\n')
-  f.write(tab + '###################################################\n')
+  format_tuple = (tab,)*11
```
Comment (forsyth2) [inline model_zoo/models/densenet/generator_densenet.py]: can  remove.

**Comment by forsyth2** [comment]
@ndryden if you have some time, do you want to review this pull request? The script `generator_densenet.py` generates the two (DenseNet 121 & 161) prototext files. Perhaps the most relevant thing to look at is the output:
```
Model 0 test objective function : 6.52246
Model 0 test categorical accuracy : 10.38%
Model 0 test top-5 categorical accuracy : 25.592%
Model 0 test run time : 118.638s
Model 0 test mini-batch time statistics : 0.605294s mean, 1.60414s max, 0.23436s min, 0.267605s stdev
```

(Output obtained from running `salloc --nodes=8 --partition=pbatch --time=600`, `export MV2_USE_CUDA=1`, and `srun build/gnu.Release.pascal.llnl.gov/lbann/build/model_zoo/lbann --reader=/usr/workspace/wsb/forsyth2/lbann/model_zoo/data_readers/data_reader_imagenet.prototext --model=/usr/workspace/wsb/forsyth2/lbann/model_zoo/models/densenet/model_densenet_121.prototext --num_epochs=5 --optimizer=/usr/workspace/wsb/forsyth2/lbann/model_zoo/optimizers/opt_sgd.prototext > /usr/workspace/wsb/forsyth2/lbann/bamboo/integration_tests/output/densenet_exe_output.txt 2> /usr/workspace/wsb/forsyth2/lbann/bamboo/integration_tests/error/densenet_exe_error.txt`)

**Comment by ndryden** [comment]
What percent of the ImageNet data was this run with?

**Comment by forsyth2** [comment]
@ndryden The results I listed were from 100% of ImageNet with 5 epochs and 8 nodes on Pascal.

**Comment by ndryden** [comment]
Results after ten epochs (on 16 Pascal nodes):
```
--------------------------------------------------------------------------------
[9] Epoch : stats formated [tr/v/te] iter/epoch = [4955/51/196]
            global MB = [ 256/ 256/ 256] global last MB = [ 132  /  11  /  80  ]
             local MB = [ 256/ 256/ 256]  local last MB = [ 132+0/  11+0/  80+0]
--------------------------------------------------------------------------------
Model 0 training epoch 9 objective function : 4.39696
Model 0 training epoch 9 categorical accuracy : 30.2296%
Model 0 training epoch 9 top-5 categorical accuracy : 55.8277%
Model 0 training epoch 9 run time : 2504.02s
Model 0 training epoch 9 mini-batch time statistics : 0.504973s mean, 2.47448s max, 0.320693s min, 0.124109s stdev
Model 0 validation objective function : 5.11686
Model 0 validation categorical accuracy : 25.3844%
Model 0 validation top-5 categorical accuracy : 49.3951%
Model 0 validation run time : 15.8051s
Model 0 validation mini-batch time statistics : 0.309896s mean, 0.595547s max, 0.157621s min, 0.0793712s stdev
Model 0 test objective function : 5.1493
Model 0 test categorical accuracy : 25.076%
Model 0 test top-5 categorical accuracy : 48.99%
Model 0 test run time : 91.2995s
Model 0 test mini-batch time statistics : 0.465808s mean, 1.49299s max, 0.258496s min, 0.18576s stdev
```

The model seems to be learning fine. I couldn't find a breakdown of per-epoch training/validation accuracy, unfortunately. My guess is it would take about three days to train to convergence, which might be the best way to compare. That might not be feasible, though.

**Comment by ndryden** [comment]
Regarding my point 3 above: We use He normal initialization by default, so you can probably disregard that one.

**Comment by forsyth2** [comment]
It looks like some of my output got overwritten, but here was the latest data I could retrieve from running `srun /usr/workspace/wsb/forsyth2/lbann/build/gnu.Release.pascal.llnl.gov/lbann/build/model_zoo/lbann --reader=/usr/workspace/wsb/forsyth2/lbann/model_zoo/data_readers/\
data_reader_imagenet.prototext --model=/usr/workspace/wsb/forsyth2/lbann/model_zoo/models/densenet/model_densenet_121.prototext --optimizer=/usr/workspace/wsb/forsyth2\
/lbann/model_zoo/optimizers/opt_sgd.prototext --learn_rate=0.1 > /usr/workspace/wsb/forsyth2/lbann/bamboo/integration_tests/output/densenet_exe_output.txt 2> /usr/wo\
rkspace/wsb/forsyth2/lbann/bamboo/integration_tests/error/densenet_exe_error.txt`:

```
--------------------------------------------------------------------------------
[65] Epoch : stats formated [tr/v/te] iter/epoch = [4955/51/196]
            global MB = [ 256/ 256/ 256] global last MB = [ 132  /  11  /  80  ]
             local MB = [ 256/ 256/ 256]  local last MB = [ 132+0/  11+0/  80+0]
--------------------------------------------------------------------------------
Model 0 training epoch 65 objective function : 1.92923
Model 0 training epoch 65 categorical accuracy : 66.5179%
Model 0 training epoch 65 top-5 categorical accuracy : 87.1335%
Model 0 training epoch 65 run time : 1471.31s
Model 0 training epoch 65 mini-batch time statistics : 0.29693s mean, 2.73797s max, 0.235493s min, 0.0855407s stdev
Checkpoint: epoch 66 step 327030 ...
Checkpoint complete: Epoch=66 Step=327030 (7.388051 secs, 40068960 bytes, 5.172236 MB/sec)
Model 0 validation objective function : 2.13146
Model 0 validation categorical accuracy : 62.7976%
Model 0 validation top-5 categorical accuracy : 84.607%
Model 0 validation run time : 6.86679s
Model 0 validation mini-batch time statistics : 0.134637s mean, 0.272208s max, 0.0769241s min, 0.043011s stdev
Checkpoint: epoch 66 step 327030 ...
Checkpoint complete: Epoch=66 Step=327030 (3.476280 secs, 51388 bytes, 0.014098 MB/sec)
--------------------------------------------------------------------------------
[66] Epoch : stats formated [tr/v/te] iter/epoch = [4955/51/196]
            global MB = [ 256/ 256/ 256] global last MB = [ 132  /  11  /  80  ]
             local MB = [ 256/ 256/ 256]  local last MB = [ 132+0/  11+0/  80+0]
--------------------------------------------------------------------------------
```
I will try running again with the output/error redirect set to append instead of overwrite.

**Comment by ndryden** [comment]
These results seem like good progress. The DenseNet paper (table 3-- LBANN uses single-crops) reports ~75% top-1 and ~92% top-5 test accuracy after 90 epochs. I would expect us to be a little less than that because we're not doing exactly the same data augmentation.

**Comment by forsyth2** [comment]
Here are my results for running 90 epochs (https://lc.llnl.gov/bamboo/browse/LBANN-FOR3-17 - `/usr/workspace/wsb/lbannusr/bamboo/lbann-pascal83-lbannusr-1/xml-data/build-dir/LBANN-FOR3-DNP/bamboo/integration_tests/output/densenet_exe_output.txt`)
```
--------------------------------------------------------------------------------
[89] Epoch : stats formated [tr/v/te] iter/epoch = [4955/51/196]
            global MB = [ 256/ 256/ 256] global last MB = [ 132  /  11  /  80  ]
             local MB = [ 256/ 256/ 256]  local last MB = [ 132+0/  11+0/  80+0]
--------------------------------------------------------------------------------
Model 0 training epoch 89 objective function : 2.40482
Model 0 training epoch 89 categorical accuracy : 56.9671%
Model 0 training epoch 89 top-5 categorical accuracy : 81.0007%
Model 0 training epoch 89 run time : 1514.5s
Model 0 training epoch 89 mini-batch time statistics : 0.305648s mean, 3.19047s max, 0.226162s min, 0.103333s stdev
Checkpoint: epoch 90 step 445950 ...
Checkpoint complete: Epoch=90 Step=445950 (2.910263 secs, 40068960 bytes, 13.130340 MB/sec)
Model 0 validation objective function : 2.57586
Model 0 validation categorical accuracy : 54.0629%
Model 0 validation top-5 categorical accuracy : 78.4092%
Model 0 validation run time : 6.99722s
Model 0 validation mini-batch time statistics : 0.137195s mean, 0.260542s max, 0.0354673s min, 0.0414623s stdev
Checkpoint: epoch 90 step 445950 ...
Checkpoint complete: Epoch=90 Step=445950 (1.115483 secs, 51388 bytes, 0.043934 MB/sec)
Model 0 test objective function : 2.66029
Model 0 test categorical accuracy : 52.762%
Model 0 test top-5 categorical accuracy : 77.58%
Model 0 test run time : 27.9414s
Model 0 test mini-batch time statistics : 0.142556s mean, 0.305858s max, 0.0777513s min, 0.0353155s stdev
```

File: model_zoo/models/densenet/generator_densenet.py [hunk 3]
```diff
@@ -70,16 +72,16 @@ def initial_layer(f, tab, num_initial_channels):
   input_name = input_layer(f, tab, name, [], io_buffer='partitioned')
   # conv_dims_i=7, conv_strides_i=2 specified in "Densely Connected Convolutional Networks"
   # conv_pads_i=3 specified in pytorch implementation
-  conv_block_name = conv_block(f, tab, name, [input_name], num_dims=2, num_output_channels=num_initial_channels, conv_dims_i=7, conv_pads_i=3, conv_strides_i=2)
+  conv_block_name = conv_block_initial(f, tab, name, [input_name], num_dims=2, num_output_channels=num_initial_channels, conv_dims_i=7, conv_pads_i=3, conv_strides_i=2)
   # pool_dims_i=3, pool_strides_i=2, pool_mode='max' specified in "Densely Connected Convolutional Networks"
   pool_name = pool(f, tab, name, [conv_block_name], num_dims=2, pool_dims_i=3, pool_strides_i=2, pool_mode='max')
```
Comment (ndryden) [inline model_zoo/models/densenet/generator_densenet.py]: This should have padding 1.

File: model_zoo/models/densenet/generator_densenet.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+  f.write('}\n')
+
+def initial_layer(f, tab, num_initial_channels):
+  name = 'initial_layer'
+  input_name = input_layer(f, tab, name, [], io_buffer='partitioned')
+  # conv_dims_i=7, conv_strides_i=2 specified in "Densely Connected Convolutional Networks"
+  # conv_pads_i=3 specified in pytorch implementation
+  conv_block_name = conv_block_initial(f, tab, name, [input_name], num_dims=2, num_output_channels=num_initial_channels, conv_dims_i=7, conv_pads_i=3, conv_strides_i=2)
+  # pool_dims_i=3, pool_strides_i=2, pool_mode='max' specified in "Densely Connected Convolutional Networks"
+  pool_name = pool(f, tab, name, [conv_block_name], num_dims=2, pool_dims_i=3, pool_strides_i=2, pool_mode='max')
```
Comment (forsyth2) [inline model_zoo/models/densenet/generator_densenet.py]: Initial pooling needs padding 1.

**Comment by forsyth2** [comment]
@ndryden I'm not seeing how the number of channels get changed between the prototext and the LBANN output.

### Details
Output of `grep -A 4 -m 6 -e "name:.*part_2_conv" model_densenet_121.prototext`:
```
    name: "dense_block_1_dense_layer_1_part_2_conv"
    data_layout: "data_parallel"
    convolution {
      num_dims: 2
      num_output_channels: 64
--
    name: "dense_block_1_dense_layer_2_part_2_conv"
    data_layout: "data_parallel"
    convolution {
      num_dims: 2
      num_output_channels: 96
--
    name: "dense_block_1_dense_layer_3_part_2_conv"
    data_layout: "data_parallel"
    convolution {
      num_dims: 2
      num_output_channels: 128
--
    name: "dense_block_1_dense_layer_4_part_2_conv"
    data_layout: "data_parallel"
    convolution {
      num_dims: 2
      num_output_channels: 160
--
    name: "dense_block_1_dense_layer_5_part_2_conv"
    data_layout: "data_parallel"
    convolution {
      num_dims: 2
      num_output_channels: 192
--
    name: "dense_block_1_dense_layer_6_part_2_conv"
    data_layout: "data_parallel"
    convolution {
      num_dims: 2
      num_output_channels: 224
```
The`num_output_channels` values match up with the following norm layers' number of input channels in the pytorch results (64, 96, 128, 160, 192, 224). However, they do not match up with the following norm layers' number of input channels when running LBANN (64, 64, 160, 288, 448, 640). The full output of `grep -m 6 "part_1_norm" densenet_exe_output.txt` is:
```
dense_block_1_dense_layer_1_part_1_norm:[batch normalization(G)] Set up a layer with input  200704 and  200704 neurons. (activations = [64c x 56w x 56h, 256s])
dense_block_1_dense_layer_2_part_1_norm:[batch normalization(G)] Set up a layer with input  200704 and  200704 neurons. (activations = [64c x 56w x 56h, 256s])
dense_block_1_dense_layer_3_part_1_norm:[batch normalization(G)] Set up a layer with input  501760 and  501760 neurons. (activations = [160c x 56w x 56h, 256s])
dense_block_1_dense_layer_4_part_1_norm:[batch normalization(G)] Set up a layer with input  903168 and  903168 neurons. (activations = [288c x 56w x 56h, 256s])
dense_block_1_dense_layer_5_part_1_norm:[batch normalization(G)] Set up a layer with input 1404928 and 1404928 neurons. (activations = [448c x 56w x 56h, 256s])
dense_block_1_dense_layer_6_part_1_norm:[batch normalization(G)] Set up a layer with input 2007040 and 2007040 neurons. (activations = [640c x 56w x 56h, 256s])
```
The only layer between the `part_2_conv` and `part_1_norm` is the concatenation layer. The only thing that can be changed about the concatenation layer is `concatenation_axis`. Setting it to 0 yields the same output as above and setting it to 1 or 2 yields only the first two lines as above (which is a bit strange since the default value seems to be 2, as indicated in https://github.com/LLNL/lbann/blob/develop/src/proto/lbann.proto#L929).

File: model_zoo/models/densenet/generator_densenet.py [hunk 5]
```diff
@@ -1,7 +1,13 @@
...
+# See https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py
 # See "Densely Connected Convolutional Networks" by Huang et. al
 
+# Note that `--learn_rate` for the sgd optimizer should be set to 0.1 for this model.
+
+import sys
+sys.path.insert(0, '..')
+import model_generator
+
 ### DenseNet Specific #########################################################
```
Comment (forsyth2) [inline model_zoo/models/densenet/generator_densenet.py]: Can remove this line now that general code is in `model_generator.py` (also, rename to `generator_model.py` to match naming structure).

**Comment by ndryden** [comment]
So, the value in concatenation (`int64 concatenation_axis = 2;`) is not a default value, but a field number (see [here](https://developers.google.com/protocol-buffers/docs/proto3#simple)). The default value is 0, which should concatenate along the channel axis, which is what we want.

I think you may be right that there's something wrong with how the concat layers are specified though. Perhaps we don't have all the right parents?

**Comment by forsyth2** [comment]
Closing in favor of #987.
